src/vsr.zig

```diff
@@ -1026,3 +998,88 @@ pub fn quorums(replica_count: u8) struct {
         .view_change = quorum_view_change,
     };
 }
+
+/// The SuperBlock's persisted VSR headers.
+/// One of the following:
+///
+/// - SV headers (consecutive chain)
+/// - DVC headers (disjoint chain)
+pub const ViewChangeHeaders = struct {
+    slice: []const Header,
+
+    pub const BoundedArray = std.BoundedArray(Header, constants.pipeline_prepare_queue_max);
+
+    pub fn init(slice: []const Header) ViewChangeHeaders {
+        ViewChangeHeaders.verify(slice);
+
+        return .{ .slice = slice };
+    }
+
+    pub fn verify(slice: []const Header) void {
+        assert(slice.len > 0);
+        assert(slice.len <= constants.pipeline_prepare_queue_max);
+
+        var child: ?*const Header = null;
+        for (slice) |*header| {
+            assert(header.valid_checksum());
+            assert(header.command == .prepare);
+
+            if (child) |child_header| {
+                assert(header.op < child_header.op);
+                assert(header.view <= child_header.view);
+                assert((header.op + 1 == child_header.op) ==
+                    (header.checksum == child_header.parent));
+                assert(header.timestamp < child_header.timestamp);
+            }
+            child = header;
+        }
+    }
+
+    const ViewRange = struct {
+        min: u32, // inclusive
+        max: u32, // inclusive
+
+        pub fn contains(range: ViewRange, view: u32) bool {
+            return range.min <= view and view <= range.max;
+        }
+    };
+
+    /// Returns the range of possible views (of prepare, not commit) for a message that is part of
+    /// the same log_view as these headers.
+    pub fn view_for_op(headers: ViewChangeHeaders, op: u64, log_view: u32) ViewRange {
+        const header_newest = &headers.slice[0];
+        const header_oldest = &headers.slice[headers.slice.len - 1];
+
+        if (op < header_oldest.op) return .{ .min = 0, .max = header_oldest.view };
+        if (op > header_newest.op) return .{ .min = log_view, .max = log_view };
```

Comment:
I'm suprised this isn't `.min = header_newest.view`. Why are we sure that ops after the headers must have been from the current view? Because we would have saved them to the superblock if they were from before log_view?

Reply:
> Why are we sure that ops after the headers must have been from the current view?

The actual headers in question here are either:
- DVC headers, or
- SV headers
for a particular `log_view`.

To answer your question for each separately:
- DVC: If we have the DVC headers for a `log_view` `V`, we must be in `view_change` status trying to transition to some view greater than `V`. So we will never prepare anything else as part of view `V`. So `header_newest.view` would actually work too (in the DVC case).
- SV: If we have the SV headers for a `log_view`, we can continue add to them (preparing more ops), but those ops will always be part of the `log_view`. ("Because we would have saved them to the superblock if they were from before log_view?" is exactly right).

Reply:
Could you add this explanation and the above one as comments?

Reply:
Rewritten to:
```
    /// Returns the range of possible views (of prepare, not commit) for a message that is part of
    /// the same log_view as these headers.
    ///
    /// - When these are DVC headers for a log_view=V, we must be in view_change status working to
    ///   transition to a view beyond V. So we will never prepare anything else as part of view V.
    /// - When these are SV headers for a log_view=V, we can continue to add to them (by preparing
    ///   more ops), but those ops will laways be part of the log_view. If they were prepared during
    ///   a view prior to the log_view, they would already be part of the headers.
```